[css-anchor-position-1] Make anchors shadow-DOM-scoped

tuankiet65 opened the pull request:
#### f205117295d5ce8eabd95aec9981b3fb1e439334
<pre>
[css-anchor-position-1] Make anchors shadow-DOM-scoped
<a href="https://rdar.apple.com/151797497">rdar://151797497</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=293382">https://bugs.webkit.org/show_bug.cgi?id=293382</a>

Reviewed by Antti Koivisto.

Right now, anchors are uniquely identified by their string identifier,
which is not enough since anchors are scoped to their shadow DOM tree.
This patch introduces ResolvedScopedName, which contains the same info
as a ScopedName, except the scope ordinal has been resolved to the
concrete Style::Scope object. ResolvedScopedName are then used
as unique identifier for anchor names. Code that uses AtomString
to identify anchors now uses ResolvedScopedName instead.

Combined changes:
* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-name-in-shadow-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-name-in-shadow-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-transition-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-transition-003-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::findAnchorForAnchorFunctionAndAttemptResolution):
(WebCore::Style::findLastAcceptableAnchorWithName):
(WebCore::Style::collectAnchorsForAnchorName):
(WebCore::Style::AnchorPositionEvaluator::findAnchorsForAnchorPositionedElement):
(WebCore::Style::AnchorPositionEvaluator::updateAnchorPositionedStateForLayoutTimePositioned):
* Source/WebCore/style/AnchorPositionEvaluator.h:
* Source/WebCore/style/ResolvedScopedName.cpp: Added.
(WebCore::Style::ResolvedScopedName::ResolvedScopedName):
(WebCore::Style::ResolvedScopedName::createFromScopedName):
* Source/WebCore/style/ResolvedScopedName.h: Added.
(WebCore::Style::ResolvedScopedName::name const):
(WebCore::Style::ResolvedScopedName::scopeIdentifier const):
(WebCore::Style::add):
(WTF::DefaultHash&lt;WebCore::Style::ResolvedScopedName&gt;::hash):
(WTF::DefaultHash&lt;WebCore::Style::ResolvedScopedName&gt;::equal):
(WTF::HashTraits&lt;WebCore::Style::ResolvedScopedName&gt;::emptyValue):
(WTF::HashTraits&lt;WebCore::Style::ResolvedScopedName&gt;::constructDeletedValue):
(WTF::HashTraits&lt;WebCore::Style::ResolvedScopedName&gt;::isDeletedValue):
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::forOrdinal):
* Source/WebCore/style/StyleScope.h:
* Source/WebCore/style/StyleScopeIdentifier.h: Added.

Canonical link: <a href="https://commits.webkit.org/295465@main">https://commits.webkit.org/295465@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0478efadde1daeb73eb8472fe5800177ba2b170e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104915 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24629 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15049 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110131 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55589 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106956 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25029 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33173 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79660 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107921 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19467 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94683 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59967 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/19218 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12761 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54972 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88932 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12807 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112605 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32080 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23596 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88737 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32444 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90909 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88368 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22591 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33262 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11036 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27395 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32005 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37430 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31797 "Built successfully") | | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35138 "Build is in progress. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33356 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Scony reviewed (commented):
LGTM.

The only thing that I started wondering about from the start is the name of the class i.e. `AnchorName` which on the one hand is natural but on the other suggests to the reader this is something simple and string-like. I thought about it for a while and it's tricky. IMO the comment describing the class may be a good hint:
> Unique identifier for an anchor in a document (...)

That would suggest something like `AnchorIdentifier`, `AnchorUUID`, `AnchorDocumentIdentifier` etc. but I'm not sure about such names either. So all in all, I'd say this may be something to brainstorm or perhaps address in the future if it ends up confusing readers.

anttijk commented on Source/WebCore/style/AnchorName.cpp:
```diff
@@ -0,0 +1,66 @@
+AnchorName AnchorName::createFromScopedName(const Element& element, const ScopedName& scopedName)
+{
+    // FIXME: this const_cast is particularly ugly.
+    const auto* scope = Scope::forOrdinal(const_cast<Element&>(element), scopedName.scopeOrdinal);
```
you can just add `static const Scope* forOrdinal(const Element&, ScopeOrdinal);` instead of commenting here

tuankiet65 replied:
Fixed.

anttijk commented on Source/WebCore/style/AnchorName.h:
```diff
@@ -0,0 +1,88 @@
+// as a string, and the Style::Scope object that the anchor is declared in to.
+// The only recommended usage is AnchorName::createFromScopedName, which takes
+// an anchor as a ScopedName, and create an AnchorName out of it.
+class AnchorName {
```
Lets call this something like `ResolvedScopedName`, there might be other clients.

tuankiet65 replied:
I renamed it to `ResolvedScopedName`

anttijk commented on Source/WebCore/style/StyleScope.h:
```diff
@@ -71,7 +73,7 @@ class Resolver;
 struct MatchResult;
 
-class Scope final : public CanMakeWeakPtr<Scope>, public CanMakeCheckedPtr<Scope> {
+class Scope final : public CanMakeWeakPtr<Scope>, public CanMakeCheckedPtr<Scope>, public Identified<ScopeIdentifier> {
```
Good idea!

tuankiet65 committed:
[css-anchor-position-1] Make anchors shadow-DOM-scoped
rdar://151797497
https://bugs.webkit.org/show_bug.cgi?id=293382

Reviewed by Antti Koivisto.

Right now, anchors are uniquely identified by their string identifier,
which is not enough since anchors are scoped to their shadow DOM tree.
This patch introduces ResolvedScopedName, which contains the same info
as a ScopedName, except the scope ordinal has been resolved to the
concrete Style::Scope object. ResolvedScopedName are then used
as unique identifier for anchor names. Code that uses AtomString
to identify anchors now uses ResolvedScopedName instead.

Combined changes:
* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-name-in-shadow-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-name-in-shadow-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-transition-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-transition-003-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::findAnchorForAnchorFunctionAndAttemptResolution):
(WebCore::Style::findLastAcceptableAnchorWithName):
(WebCore::Style::collectAnchorsForAnchorName):
(WebCore::Style::AnchorPositionEvaluator::findAnchorsForAnchorPositionedElement):
(WebCore::Style::AnchorPositionEvaluator::updateAnchorPositionedStateForLayoutTimePositioned):
* Source/WebCore/style/AnchorPositionEvaluator.h:
* Source/WebCore/style/ResolvedScopedName.cpp: Added.
(WebCore::Style::ResolvedScopedName::ResolvedScopedName):
(WebCore::Style::ResolvedScopedName::createFromScopedName):
* Source/WebCore/style/ResolvedScopedName.h: Added.
(WebCore::Style::ResolvedScopedName::name const):
(WebCore::Style::ResolvedScopedName::scopeIdentifier const):
(WebCore::Style::add):
(WTF::DefaultHash<WebCore::Style::ResolvedScopedName>::hash):
(WTF::DefaultHash<WebCore::Style::ResolvedScopedName>::equal):
(WTF::HashTraits<WebCore::Style::ResolvedScopedName>::emptyValue):
(WTF::HashTraits<WebCore::Style::ResolvedScopedName>::constructDeletedValue):
(WTF::HashTraits<WebCore::Style::ResolvedScopedName>::isDeletedValue):
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::forOrdinal):
* Source/WebCore/style/StyleScope.h:
* Source/WebCore/style/StyleScopeIdentifier.h: Added.

Canonical link: https://commits.webkit.org/295465@main
